Add Check Support for Java 21 Unnamed Variables & Patterns Syntax: LambdaParameterName

child of https://github.com/checkstyle/checkstyle/issues/14942
I have read check documentation: https://checkstyle.org/checks/naming/lambdaparametername.html#LambdaParameterName
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
PS D:\CS\test> cat src/Test.java
import java.util.List;

public class Test {
        List<String> list = List.of("a", "b", "c");
        List<String> newList = list.stream().map(_ -> "x").toList(); // violation
}
PS D:\CS\test> cat config.xml   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
         <module name="LambdaParameterName"/>
    </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java              
Starting audit...
[ERROR] D:\CS\test\src\Test.java:5:50: Name '_' must match pattern '^[a-z][a-zA-Z0-9]*$'. [LambdaParameterName]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 
```


Describe what you expect in detail.

`LambdaParameterName` should allow unnamed pattern variables by default

--- a/src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionLambdaParameterNameTest.java
+++ b/src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionLambdaParameterNameTest.java
@@ -46,7 +46,7 @@ public class XpathRegressionLambdaParameterNameTest extends AbstractXpathTestSup
 
         final DefaultConfiguration moduleConfig =
                 createModuleConfig(LambdaParameterNameCheck.class);
-        final String defaultPattern = "^[a-z][a-zA-Z0-9]*$";
+        final String defaultPattern = "^([a-z][a-zA-Z0-9]*|_)$";
 
         final String[] expectedViolation = {
             "7:44: " + getCheckMessage(LambdaParameterNameCheck.class,

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LambdaParameterNameCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/LambdaParameterNameCheck.java
@@ -33,7 +33,7 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * <li>
  * Property {@code format} - Sets the pattern to match valid identifiers.
  * Type is {@code java.util.regex.Pattern}.
- * Default value is {@code "^[a-z][a-zA-Z0-9]*$"}.
+ * Default value is {@code "^([a-z][a-zA-Z0-9]*|_)$"}.
  * </li>
  * </ul>
  * <p>
@@ -54,7 +54,7 @@ public class LambdaParameterNameCheck extends AbstractNameCheck {
 
     /** Creates new instance of {@code LambdaParameterNameCheck}. */
     public LambdaParameterNameCheck() {
-        super("^[a-z][a-zA-Z0-9]*$");
+        super("^([a-z][a-zA-Z0-9]*|_)$");
     }
 
     @Override

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/naming/LambdaParameterNameCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/naming/LambdaParameterNameCheck.xml
@@ -8,7 +8,7 @@
  Checks lambda parameter names.
  &lt;/p&gt;</description>
          <properties>
-            <property default-value="^[a-z][a-zA-Z0-9]*$"
+            <property default-value="^([a-z][a-zA-Z0-9]*|_)$"
                       name="format"
                       type="java.util.regex.Pattern">
                <description>Sets the pattern to match valid identifiers.</description>

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LambdaParameterNameCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/naming/LambdaParameterNameCheckTest.java
@@ -79,7 +79,7 @@ public class LambdaParameterNameCheckTest extends AbstractModuleTestSupport {
     @Test
     public void testLambdaParameterNameSwitchExpression() throws Exception {
 
-        final String pattern = "^[a-z][a-zA-Z0-9]*$";
+        final String pattern = "^([a-z][a-zA-Z0-9]*|_)$";
 
         final String[] expected = {
             "19:35: " + getCheckMessage(MSG_INVALID_PATTERN, "Word", pattern),
@@ -91,6 +91,22 @@ public class LambdaParameterNameCheckTest extends AbstractModuleTestSupport {
 
         verifyWithInlineConfigParser(
                 getNonCompilablePath("InputLambdaParameterNameSwitchExpression.java"),
+            expected);
+    }
+
+    @Test
+    public void testLambdaParameterNameUnnamed() throws Exception {
+
+        final String pattern = "^([a-z][a-zA-Z0-9]*|_)$";
+
+        final String[] expected = {
+            "30:36: " + getCheckMessage(MSG_INVALID_PATTERN, "__", pattern),
+            "34:36: " + getCheckMessage(MSG_INVALID_PATTERN, "_BAD", pattern),
+            "37:36: " + getCheckMessage(MSG_INVALID_PATTERN, "BAD_", pattern),
+        };
+
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputLambdaParameterNameUnnamed.java"),
             expected);
     }
 

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/lambdaparametername/InputLambdaParameterNameSwitchExpression.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/lambdaparametername/InputLambdaParameterNameSwitchExpression.java
@@ -1,6 +1,6 @@
 /*
 LambdaParameterName
-format = (default)^[a-z][a-zA-Z0-9]*$
+format = (default)^([a-z][a-zA-Z0-9]*|_)$
 
 
 */

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/lambdaparametername/InputLambdaParameterNameUnnamed.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/lambdaparametername/InputLambdaParameterNameUnnamed.java
@@ -1,0 +1,47 @@
+/*
+LambdaParameterName
+format = (default)^([a-z][a-zA-Z0-9]*|_)$
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.naming.lambdaparametername;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class InputLambdaParameterNameUnnamed {
+
+    void method(Object o) {
+        List<Integer> numbers = Arrays.asList(1, 2, 3);
+        numbers = numbers.stream()
+                .map(_ -> 0)
+                .toList();
+        List<String> strings = Arrays.asList("a", "b", "c");
+
+        System.out.println(
+                strings.stream().collect(Collectors.toMap(String::toUpperCase,
+                                   _ -> "NODATA"))); // ok, unnamed lambda parameter
+
+        System.out.println(
+                strings.stream().collect(Collectors.toMap(String::toUpperCase,
+                                   __ -> "NODATA"))); // violation, 'Name '__' must match.*'
+
+        System.out.println(
+                strings.stream().collect(Collectors.toMap(String::toUpperCase,
+                                   _BAD -> "NODATA"))); // violation, 'Name '_BAD' must match.*'
+        System.out.println(
+                strings.stream().collect(Collectors.toMap(String::toUpperCase,
+                                   BAD_ -> "NODATA"))); // violation, 'Name 'BAD_' must match.*'
+
+        switch (o) {
+            case Integer __ -> {} // ok, this is pattern variable not a lambda parameter
+            case String _ -> {}
+            default -> {}
+        }
+
+    }
+
+}

--- a/src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/naming/LambdaParameterNameCheckExamplesTest.java
+++ b/src/xdocs-examples/java/com/puppycrawl/tools/checkstyle/checks/naming/LambdaParameterNameCheckExamplesTest.java
@@ -34,7 +34,7 @@ public class LambdaParameterNameCheckExamplesTest extends AbstractExamplesModule
     @Test
     public void testExample1() throws Exception {
         final String[] expected = {
-            "17:11: " + getCheckMessage(MSG_INVALID_PATTERN, "S", "^[a-z][a-zA-Z0-9]*$"),
+            "17:11: " + getCheckMessage(MSG_INVALID_PATTERN, "S", "^([a-z][a-zA-Z0-9]*|_)$"),
         };
 
         verifyWithInlineConfigParser(getPath("Example1.java"), expected);

--- a/src/xdocs/checks/naming/lambdaparametername.xml
+++ b/src/xdocs/checks/naming/lambdaparametername.xml
@@ -28,7 +28,7 @@
               <td>format</td>
               <td>Sets the pattern to match valid identifiers.</td>
               <td><a href="../../property_types.html#Pattern">Pattern</a></td>
-              <td><code>&quot;^[a-z][a-zA-Z0-9]*$&quot;</code></td>
+              <td><code>&quot;^([a-z][a-zA-Z0-9]*|_)$&quot;</code></td>
               <td>8.11</td>
             </tr>
           </table>

